### PR TITLE
Await on mismatched future types

### LINKED ISSUE #61076: suggest `await` for future-related type errors
A common problem when using async-await is to forget to invoke `.await`. This often shows up as a type error, e.g. in a case like this ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ef1caa955ef1aceadf33aadd348e646c)):

```rust
#![feature(async_await)]

async fn make_u32() -> u32 {
    22
}

fn take_u32(x: u32) {

}

async fn foo() {
    let x = make_u32();
    take_u32(x);
}

fn main() { }
```

currently we give a rather generic error:

```
error[E0308]: mismatched types
  --> src/main.rs:13:14
   |
13 |     take_u32(x);
   |              ^ expected u32, found opaque type
   |
   = note: expected type `u32`
              found type `impl std::future::Future`
```

it'd be nice if we could suggest adding `.await` somewhere.

### PR DESCRIPTION
**Pull request by csmoe**
Closes #61076
This PR suggests to `await` on:
1. `async_fn().bar() => async_fn().await.bar()`
2. `async_fn().field => async_fn().await.field`
3. ` if let x = async() {} => if let x = async().await {}`

r? @tmandry @estebank

### CODE REVIEW

File: src/test/ui/async-await/issue-61076.stderr [hunk 1]
```diff
@@ -22,6 +22,46 @@ LL |     t?;
...
+   |                  |
+   |                  help: consider await before field access: `tuple().await.0`
+
+error[E0609]: no field `a` on type `impl std::future::Future`
+  --> $DIR/issue-61076.rs:60:28
+   |
+LL |     let _: i32 = struct_().a;
+   |                  ----------^
+   |                  |
+   |                  help: consider await before field access: `struct_().await.a`
```
Comment (estebank) [inline src/test/ui/async-await/issue-61076.stderr:39]: Looking at this span makes me think that we might be suggesting the following resulting code: `struct_().await.aa`

File: src/librustc_infer/infer/error_reporting/mod.rs [hunk 2]
```diff
@@ -1547,6 +1551,74 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                            exp_span
+                        };
+                        if let Ok(ty_span) = self.tcx.sess.source_map().span_to_snippet(span) {
+                            diag.span_suggestion(
+                                span,
+                                "consider await on this",
+                                format!("{}.await", ty_span),
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
```
Comment (estebank) [inline src/librustc_infer/infer/error_reporting/mod.rs:1615]:
```suggestion
                            diag.span_suggestion_verbose(
                                ty_span.shrink_to_hi(),
                                "consider awaiting on the future",
                                ".await".to_string(),
                                Applicability::MaybeIncorrect,
                            );
                        }
```

**Comment by estebank** [comment]
@csmoe sorry, this fell through the cracks of my queue. Love the changes. Left some nitpicks but otherwise it LGTM!

File: src/librustc_typeck/check/expr.rs [hunk 3]
```diff
@@ -1509,13 +1509,73 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        let mut obligations = vec![];
+        if let Some(projection_ty) = projection_ty {
+            let normalized_ty = rustc_trait_selection::traits::normalize_projection_type(
+                &mut selcx,
+                param_env,
+                projection_ty,
+                cause,
+                0,
+                &mut obligations,
+            );
```
Comment (tmandry) [inline src/librustc_typeck/check/expr.rs:1549]: Is there a central place somewhere that we could define a helper to do this? I'd rather not duplicate it in all these places.

(If we can't extract the normalizing code that's fine, I'd prefer to, but we could just extract the parts before it.)

Comment (estebank) [reply src/librustc_typeck/check/expr.rs:1549]: Having a `src/librustc_typeck/check/diagnostics.rs` would make sense, and would follow the pattern in other parts of the compiler.

I made a quick search for `SelectionContext::new(` as I would imagine that any code doing this same thing would be using it, but I couldn't find any _diagnostics_ use in `typeck`.

**Comment by csmoe** [comment]
@tmandry @estebank I query'fied the loop to finding `Future::Output` from predicates.

**Comment by estebank** [comment]
These suggestions became dead code at some point (😩), but will be emitted again after https://github.com/rust-lang/rust/pull/78214.

**Pull request merged**